### PR TITLE
[R4R]Implementation of SHA3-256 FIPS 202 hash precompile and EcRecover Uncompressed public key precompile to integrate with ICON blockchain

### PR DESCRIPTION
**Pull request by simsonraj**
### Description

- Implementation of SHA3-256 FIPS 202 hash precompile to integrate with ICON blockchain
- EcRecover Uncompressed public key precompile to integrate with ICON blockchain

### Rationale

- In order to validate the merkle tree data from ICON, the SHA3-256 hashing function is required because all of ICON blockchain's hash and transaction hash are created using the SHA3-256 hashing function
- It is needed to validate signatures of validators in ICON. Currently ICON uses a similar address building scheme to ethereum. ICON uses a part of the hash value of the public key. But we are using a different hash function, SHA3-256 FIPS.

### Example

Test cases done to check the input and output of the functions for the given parameters below

sha3fips
params
data : ‘0x0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’
returns : ‘0xc7647f7e251bf1bd70863c8693e93a4e77dd0c9a689073e987d51254317dc704’

ecrecoverPublicKey
params
hash : ‘0xc5d6c454e4d7a8e8a654f5ef96e8efe41d21a65b171b298925414aa3dc061e37’
v : ‘0x00’
r : ‘0x4011de30c04302a2352400df3d1459d6d8799580dceb259f45db1d99243a8d0c’
s : ‘0x64f548b7776cb93e37579b830fc3efce41e12e0958cda9f8c5fcad682c610795’
returns : ‘0x0448250ebe88d77e0a12bcf530fe6a2cf1ac176945638d309b840d631940c93b78c2bd6d16f227a8877e3f1604cd75b9c5a8ab0cac95174a8a0a0f8ea9e4c10bca’

### Changes

Notable changes: 
* Added the new methods on to the contracts.go file 
* And the related testcases on to the contracts_test.go file

### Related issues

[Precompiles_test_sol.txt](https://github.com/binance-chain/bsc/files/6952664/Precompiles_test_sol.txt)

### CODE REVIEW

**Comment by mohamedelshami** [comment]
Hi, @karalabe @yutianwu would it be possible to get your help reviewing the above PR? this has been outstanding for a few months now. The change will enable interoperability for the BTP bridge for BSC, please find BEP requested by BSC team here (https://github.com/binance-chain/BEPs/pull/110).

I appreciate the BSC core dev team can be very busy with other priorities, but also so far we haven't heard any objection either on this PR. We have now completed the project but are held back by this change request.

Thank you.

**Comment by guagualvcha** [comment]
hi, teams. I will start looking into this PR now.

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:86]: Let's use `Hilbert` as the hardfork name.

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:88]: We should have another variable for this hardfork, like `PrecompiledContracts Hilbert `

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:1070]: Why encode to hex and then decode?

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:1085]: any reference code for this part? or you write from scratch?

Comment (simsonraj) [reply core/vm/contracts.go:1096]: This method is the same code from existing `ecrecover` without the the last line to return the whole publicKey

File: core/vm/contracts.go
Comment (simsonraj) [reply core/vm/contracts.go]: Thanks for pointing out, removed this, must have been a residue from testing.

File: core/vm/contracts.go
Comment (simsonraj) [reply core/vm/contracts.go]: Updated this to have another hard fork as Hilbert.
along with this fix, rebased to latest and with some fixes, moved the testing data to the json file test data

File: core/vm/contracts.go
Comment (simsonraj) [reply core/vm/contracts.go]: @guagualvcha Changed this with other fixes. Please help review the latest.

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:127]: `common.BytesToAddress([]byte{5}): &bigModExp{eip2565: true}` ==> `	common.BytesToAddress([]byte{5}):&bigModExp{},`

and add 
```
common.BytesToAddress([]byte{100}): &tmHeaderValidate{},
common.BytesToAddress([]byte{101}): &iavlMerkleProofValidate{},
```

Comment (guagualvcha) [reply core/vm/contracts.go:127]: sorry, even you add `PrecompiledContractsHilbert` here, but seems like you did not use it, can you refer to how we use `PrecompiledContractsIstanbul`. Otherwise, the newly added native contracts won't take effect.

File: core/vm/contracts.go
Comment (guagualvcha) [inline core/vm/contracts.go:1102]: for ecrecover, it is `v := input[63] - 27`, what is the difference?

Comment (simsonraj) [reply core/vm/contracts.go:1102]: v is the last byte of the signature, and is either 27 (0x1b) or 28 (0x1c). 
this is used to recover the points on elliptic curve.

```
In most implementations, the v is just 0 or 1 internally, but 27 was added as arbitrary number for signing Bitcoin messages and Ethereum adapted that as well.
```

File: core/vm/contracts.go
Comment (simsonraj) [reply core/vm/contracts.go:127]: I have added changes to include Hilbert precompiles, I tested by running an actual solidity test. however, I'm unsure if I did make all the necessary changes there are a few things that I did not understand when I referred to Istanbul such as the Instruction set/ tx pool changes that Istanbul has, I assume it's irrelevant for this case, and ignored it.
Also, I set `nil` to the block numbers in all the chain config. assuming you will change it when you move it to testnet / mainnet
Please do let me know if I had missed anything. Thank you.

**Comment by simsonraj** [comment]
Hi @guagualvcha, Please let me know if I missed something, I'll update it ASAP. We are waiting for this merge to move BTP forward. would appreciate your timely help to take this forward.

**Comment by SpartanNode** [comment]
Hey all, been following this for a while now. We really would like to see some progress and appreciate all of you and we'd really like to see BTP moved forward.  Thank you and kind regards for letting us bump this. BTP is important for our ecosystem.

**Comment by neij** [comment]
Hi all, please world need BTP, could somebody check it and make some proggress, everybody from Icon will be grateful.

**Comment by tiamatth** [comment]
Hey there, it would be great to get some more progress on this development.

**Comment by nicholaspalacios** [comment]
Reviews still pending CC: @yutianwu @j75689

File: core/vm/contracts.go
Comment (Lizy87) [inline core/vm/contracts.go:40]: - [ ]

**Review by Lizy87 — APPROVED** [review]

**Review by Lizy87 — APPROVED** [review]

**Review by 447311011 — APPROVED** [review]

**Comment by zzzckck** [comment]
I don't think BSC will support SHA3-256 in short term, especially when there is no progress on Ethereum, we'd better to keep EVM compatible.

**Pull request closed without merging**
